### PR TITLE
feat(cache node): add cache node step1.

### PR DESCRIPTION
**Pull request by IRONICBo**
A cluster topology mechanism has been established, utilizing a HashRing for node mapping and load balancing. To promptly respond to client RPC requests when requests hit a specific node, it is essential to hold hot data in memory to minimize I/O latency.

### Overview

A Cache Manager on the cache node manages memory usage and receives file flush requests. During data eviction, concurrency issues are temporarily ignored. It is assumed that each cache node has sufficient memory, and there is no current plan to incorporate multi-tier caching for expanded storage at the node level. The in-memory cache also maintains a block cache from the disk to facilitate file reading and writing.

Only block-level read and write operations are considered, and it is suggested to avoid dispersing block read-write concurrency control, focusing instead on handling it at the RPC layer for easier debugging.

### Design

The design of our cache node is similar to that of the Storage Backend, primarily maintaining an in-memory cache. Additionally, blocks obtained from S3 are flushed to our disk, maintaining only hot data in memory. Key components considered in implementing the cache node include:

#### Key Design
- Global uniqueness is required for keys within the cluster at the block level, directly inheriting globally unique file information.
- Each block is uniquely identified with an "inum" (allocated via ETCD) and a version number indicating the block's modification time (generated during write operations using a snowflake algorithm for local ID generation).

#### Metadata Structure

```rs
struct MetaData {
    inum: u64,           // Globally unique file identifier, allocated in ETCD
    version: i64,        // Block write time, generated with snowflake
    offset: u64,         // Offset within the file
    size: u64,           // Block size
}
```

#### Block Manager Design

The Block Manager's primary role is to manage blocks on the cache node, offering top-level block read and write capabilities.

```rs
struct BlockManager {
    cache: CacheManager<MetaData, LRUPolicy<MetaData>>,
    metasets: Arc<RwLock<HashSet<MetaData>>>,
    backend: Arc<dyn Backend>,
}
```

##### Key operations:

Read: Retrieves blocks by block ID, considering version consistency.
Write: Updates or adds new blocks based on block ID.
Invalid: Removes blocks from both memory and disk cache upon request.

#### Cache Manager Design

Similar to storage management, the Cache Manager updates in-memory flags during read/write operations and relies on a policy to limit cache size.

```rs
struct CacheManager<K, P> where K: Eq + Hash + Clone, P: EvictPolicy<K> {
    policy: P,
    cache: HashMap<K, Arc<RwLock<Block>>>,
}
```

Operations:

Read: First checks the in-memory cache, then local metadata, and if absent, fetches from S3.
Write: Currently not implemented; LRU policy handles direct invalidation in memory.
Invalid: Handles requests to invalidate data in both in-memory and disk storage.

#### Policy Design

Eviction policies are interchangeable, with the default being LRU. The policy can also be based on access frequency, such as LFU, managing data in CacheManager and executing eviction when exceeding capacity limits.

Design doc: https://datenlord.feishu.cn/docx/F8N0dK50pohtAJxMeQ9cRN8KnEd

### CODE REVIEW

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -45,7 +45,8 @@ jobs:
       - name: Print Integration Test Logs
         run: cat datenlord_test.log
       - name: Upload coverage to Codecov
-        uses: codecov/codecov-action@v3
+        uses: codecov/codecov-action@v4
         with:
+          token: ${{ secrets.CODECOV_TOKEN }}
           files: codecov.info
-          fail_ci_if_error: true
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply .github/workflows/coverage.yml:51]: I have check it, this config is from xline repo, may be it is necessary for us.

File: src/cache_node/snowflake.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        self.last_timestamp = timestamp;
+        ((timestamp - TWEPOCH) << TIMESTAMP_LEFT_SHIFT)
+            | (self.datacenter_id << DATACENTER_ID_SHIFT)
+            | (self.worker_id << WORKER_ID_SHIFT)
+            | self.sequence
+    }
+
+    fn til_next_millis(&self, last_timestamp: i64) -> i64 {
+        let mut timestamp = self.time_gen();
+        while timestamp <= last_timestamp {
```
Comment (frostyplanet) [inline src/cache_node/snowflake.rs:87]: do not dead loop on cpu,  at least sleep a little

**Comment by frostyplanet** [comment]
Should rebase the conflict

**Comment by IRONICBo** [comment]
@frostyplanet Please review the latest changes.

**Review by frostyplanet — APPROVED** [review]
LGTM

**Pull request merged**